File: UserGuideCode/Fortran_code/read_con2zn_genrl_str.f
      program read_con2zn_genrl_str
c
c   Opens an existing CGNS file that contains a simple 3-D 
c   structured grid (2 zones) plus 1-to-1 connectivity 
c   information (written in GENERAL form), and reads the 
c   connectivity info.
c
c   The CGNS grid file 'grid.cgns' must already exist
c   (created using write_grid2zn_str.f plus write_con2zn_genrl_str.f)
c
c   This program uses the fortran convention that all
c   variables beginning with the letters i-n are integers,
c   by default, and all others are real
c
c   Example compilation for this program is (change paths!):
c
c   ifort -I ../CGNS_CVS/cgnslib_2.5 -c read_con2zn_genrl_str.f
c   ifort -o read_con2zn_genrl_str read_con2zn_genrl_str.o -L ../CGNS_CVS/cgnslib_2.5/LINUX -lcgns
c
c   (../CGNS_CVS/cgnslib_2.5/LINUX/ is the location where the compiled
c   library libcgns.a is located)
c
c   cgnslib_f.h file must be located in directory specified by -I during compile:
      include 'cgnslib_f.h'
c
      parameter (maxpnts=400)
      dimension ipnts(3,maxpnts),ipntsdonor(3,maxpnts)
      character donorname*32,connectname*32
c
c  READ GENERAL CONNECTIVITY INFORMATION FROM EXISTING CGNS FILE
c  open CGNS file for read-only
      call cg_open_f('grid.cgns',CG_MODE_READ,index_file,ier)
      if (ier .ne. CG_OK) call cg_error_exit_f
c  we know there is only one base (real working code would check!)
      index_base=1
c   get number of zones (should be 2 for our case)
      call cg_nzones_f(index_file,index_base,nzone,ier)
      if (nzone .ne. 2) then
         write(6,'('' Error.  This program expects 2 zones. '',i5,
     +    '' read'')') nzone
         stop
      end if
c   loop over zones 
      do index_zone=1,nzone
c   find out how many general interfaces there are in this zone
c   (for this program, there should only be one)
        call cg_nconns_f(index_file,index_base,index_zone,nconns,ier)
        if (nconns .ne. 1) then
          write(6,'('' Error.  Expecting one general interface.'',
     +      i6,'' read'')') nconns
          stop
        end if
        index_conn=nconns
c   read general connectivity info
        call cg_conn_info_f(index_file,index_base,index_zone,index_conn,
     +     connectname,location,iconnect_type,iptset_type,npts,
     +     donorname,idonor_zonetype,idonor_ptset_type,idonor_datatype,
     +     ndata_donor,ier)
        if (npts .gt. maxpnts) then
          write(6,'('' Error.  Must increase maxpnts to at least '',
     +     i5)') npts
          stop
        end if
        call cg_conn_read_f(index_file,index_base,index_zone,index_conn,
     +    ipnts,idonor_datatype,ipntsdonor,ier)
        write(6,'('' In zone '',i5,'':'')') index_zone
        write(6,'(''    donor name='',a32)') donorname
        write(6,'(''    number of connectivity pts='',i6)') npts
        write(6,'(''    grid location='',a32)') 
     +   GridLocationName(location)
        write(6,'(''    connectivity type='',a32)') 
     +   GridConnectivityTypeName(iconnect_type)
        write(6,'(''    pointset type='',a32)') 
     +   PointSetTypeName(iptset_type)
        write(6,'(''    donor zonetype='',a32)') 
     +   ZoneTypeName(idonor_zonetype)
        write(6,'(''    donor pointset type='',a32)') 
     +   PointSetTypeName(idonor_ptset_type)
        write(6,'(''    data type='',a32)') 
     +   DataTypeName(idonor_datatype)
        write(6,'(''    ipnts and ipntsdonor arrays read, only some'',
     +   '' written out here:'')')
        do i=1,10
          write(6,'('' ipnts(1,'',i2,''), (2,'',i2,''), (3,'',i2,'')='',
     +      3i4,''   ipntsdonor(1,'',i2,''), (2,'',i2,''), (3,'',i2,
     +      '')='',3i4)') i,i,i,ipnts(1,i),ipnts(2,i),ipnts(3,i),
     +      i,i,i,ipntsdonor(1,i),ipntsdonor(2,i),ipntsdonor(3,i)
        enddo
      enddo
c  close CGNS file
      call cg_close_f(index_file,ier)
      write(6,'('' Successfully read general 1-to-1 connectivity'',
     + '' info from file grid.cgns'')')
      stop
      end

File: src/Test_UserGuideCode/C_code/read_con2zn_genrl_str.c
/*   Program read_con2zn_genrl_str   */
/*
Opens an existing CGNS file that contains a simple 3-D
structured grid (2 zones) plus 1-to-1 connectivity
information (written in GENERAL form), and reads the
connectivity info.

The CGNS grid file 'grid_c.cgns' must already exist
(created using write_grid2zn_str.c plus write_con2zn_genrl_str.c)

Example compilation for this program is (change paths if needed!):

cc -I ../.. -c read_con2zn_genrl_str.c
cc -o read_con2zn_genrl_str_c read_con2zn_genrl_str.o -L ../../lib -lcgns

(../../lib is the location where the compiled
library libcgns.a is located)
*/

#include <stdio.h>
/* cgnslib.h file must be located in directory specified by -I during compile: */
#include "cgnslib.h"

#if CGNS_VERSION < 3100
# define cgsize_t int
#endif

#define maxpnts 400

int main()
{
    int i,index_file,index_base,nzone,index_zone,nconns,index_conn;
    char donorname[33],connectname[33];
    CGNS_ENUMT(GridLocation_t) location;
    CGNS_ENUMT(GridConnectivityType_t) iconnect_type;
    CGNS_ENUMT(PointSetType_t) iptset_type,idonor_ptset_type;
    CGNS_ENUMT(ZoneType_t) idonor_zonetype;
    CGNS_ENUMT(DataType_t) idonor_datatype;
    cgsize_t npts,ndata_donor;
    cgsize_t ipnts[maxpnts][3],ipntsdonor[maxpnts][3];

/* READ GENERAL CONNECTIVITY INFORMATION FROM EXISTING CGNS FILE */
/* open CGNS file for read-only */
    if (cg_open("grid_c.cgns",CG_MODE_READ,&index_file)) cg_error_exit();
/* we know there is only one base (real working code would check!) */
    index_base=1;
/* get number of zones (should be 2 for our case) */
    cg_nzones(index_file,index_base,&nzone);
    if (nzone != 2)
    {
      printf("\nError. This program expects 2 zones. %d read.\n",nzone);
      return 1;
    }
/* loop over zones */
    for (index_zone=1; index_zone <= nzone; ++index_zone)
    {
/* find out how many general interfaces there are in this zone */
/* (for this program, there should only be one) */
      cg_nconns(index_file,index_base,index_zone,&nconns);
      if (nconns != 1)
      {
        printf("\nError.  Expecting one general interface. %i read\n",nconns);
        return 1;
      }
      index_conn=nconns;
/* read general connectivity info */
      cg_conn_info(index_file,index_base,index_zone,index_conn,connectname,&location,
                   &iconnect_type,&iptset_type,&npts,donorname,&idonor_zonetype,
                   &idonor_ptset_type,&idonor_datatype,&ndata_donor);
      if (npts > maxpnts)
      {
        printf("\nError.  Must increase maxpnts to at least %i\n",(int)npts);
        return 1;
      }
      cg_conn_read(index_file,index_base,index_zone,index_conn,ipnts[0],
                   idonor_datatype,ipntsdonor[0]);
      printf("\nIn zone %i:\n",index_zone);
      printf("   donor name=%s\n",donorname);
      printf("   number of connectivity pts=%i\n",(int)npts);
      printf("   grid location=%s\n",GridLocationName[location]);
      printf("   connectivity type=%s\n",GridConnectivityTypeName[iconnect_type]);
      printf("   pointset type=%s\n",PointSetTypeName[iptset_type]);
      printf("   donor zonetype=%s\n",ZoneTypeName[idonor_zonetype]);
      printf("   donor pointset type=%s\n",PointSetTypeName[idonor_ptset_type]);
/*    printf("   data type=%s\n",DataTypeName[idonor_datatype]); */
      printf("   ipnts and ipntsdonor arrays read, only some written out here:\n");
      for (i=0; i < 10; i++)
      {
        printf("    ipnts[%i][0], [%i][1], [%i][2]=%i,%i,%i",
               i,i,i,(int)ipnts[i][0],(int)ipnts[i][1],(int)ipnts[i][2]);
        printf("    ipntsdonor[%i][0], [%i][1], [%i][2]=%i,%i,%i\n",
               i,i,i,(int)ipntsdonor[i][0],(int)ipntsdonor[i][1],(int)ipntsdonor[i][2]);
      }
    }
/* close CGNS file */
    cg_close(index_file);
    printf("\nSuccessfully read general 1-to-1 connectivity info from file grid_c.cgns\n");
    return 0;
}
